### PR TITLE
LUN-400: Image for own vocabulary

### PR DESCRIPTION
**Pull request by SteffiStuffel**
This pull request belongs to an issue on our [bugtracker](https://issues.tuerantuer.org/secure/RapidBoard.jspa?rapidView=27&view=detail).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword _LUN_.

Should work for Android and iOS. Saves images in app specific storage for both.

### CODE REVIEW

File: android/app/src/main/java/app/lunes/MainApplication.java [hunk 1]
```diff
@@ -4,6 +4,7 @@
 import android.content.Context;
```
Comment (SteffiStuffel) [inline android/app/src/main/java/app/lunes/MainApplication.java:4]: renamed this folder, as the path for saving files to app specific storage on android will equal this.

File: src/components/CameraOverlay.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
+import React, { ReactElement, useEffect, useRef, useState } from 'react'
```
Comment (SteffiStuffel) [inline src/components/CameraOverlay.tsx:1]: Moved Camera stuff from previous QRCodeOverlay to this shared component, which is used in QRCodeOverlay and ImageSelectionOverlay.

File: src/components/__tests__/CameraOverlay.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import { fireEvent } from '@testing-library/react-native'
```
Comment (SteffiStuffel) [inline src/components/__tests__/CameraOverlay.spec.tsx:1]: Only moved from QRCodeOverlay.

**Review by charludo — COMMENTED** [review]
Looks like the failing tests are due to the same error we've talked about in `Lun-407` recently.

Adding `jest.mock('@react-navigation/native')` [here](https://github.com/digitalfabrik/lunes-app/blob/LUN-400-image-for-own-vocabulary/src/routes/user-vocabulary-list/__tests__/UserVocabularyListScreen.spec.tsx#L13) should fix it.

**Review by steffenkleinle — CHANGES_REQUESTED** [review]
Looks really good already, well done! Tested on android.

Things I noticed:
1. The error messages of the image overlay are not adjusted, e.g. it says `Kamerazugriff zum Scannen von QR-Codes erforderlich` even though I am trying to add images for vocabulary.
2. There is no kind of feedback that I already took an image. I think it would be better if the camera overlay would close after taking one image and the user has to klick on `Bild hochladen` again if more than one images are wanted (I think it will not be that common)
3. `Bild hochladen` sounds like it is uploaded to a server or something which I personally as a privacy oriented user would be very hesitant to do. Instead, the naming `Bild hinzufügen` would be better imo. Not sure if we should use plural.
4. Is it planned to add an option to add images from the gallery instead of having to take new pictures?
5. The view for user vocabulary is broken, see screenshot (probably not done in your PR though).

File: src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+interface Props {
+  setVisible: (visible: boolean) => void
+  numberOfImages: number
+  pushImage: (imageUri: string) => void
+}
+
+// TODO
+// adjust and test on ios
+// write test
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx:44]: yes

File: src/services/helpers.ts [hunk 5]
```diff
@@ -201,3 +202,14 @@ export const getSortedAndFilteredDocuments = (documents: Document[] | null, sear
...
 export const willNextExerciseUnlock = (previousScore: number | undefined, score: number): boolean =>
   score > SCORE_THRESHOLD_UNLOCK && (previousScore ?? 0) <= SCORE_THRESHOLD_UNLOCK
+
+export const getImages = (item: Document): Promise<Images> =>
+  Promise.all(
+    item.document_image.map(async image => ({
+      id: image.id,
+      image: await readFile(image.image).catch(err => {
+        reportError(err)
+        return ''
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/services/helpers.ts:212]: it was easier, but you are right, it is better to remove invalid values.

File: src/routes/process-user-vocabulary/components/Thumbnail.tsx [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+`
+
+const CloseIcon = styled(PressableOpacity)`
+  position: absolute;
+  right: 0px;
+  padding: ${props => props.theme.spacings.xxs};
+`
+
+const Image = styled(RNImage)`
+  width: ${hp('12%')}px;
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/routes/process-user-vocabulary/components/Thumbnail.tsx:20]: We use always `hp`, apart from exceptional cases.

File: src/routes/process-user-vocabulary/components/Thumbnail.tsx [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  height: ${hp('12%')}px;
+`
+
+interface ThumbnailProps {
+  image: string
+  deleteImage: () => void
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/routes/process-user-vocabulary/components/Thumbnail.tsx:26]: Not sure if I get your point, but yes there is a thumbnail, and yes it is possible to delete the image with this thumbnail. See designs: https://app.zeplin.io/project/5fc7c98a435298126bcfd4ee/screen/63074ed84e561380e4192f11

I think the naming is fine.

File: src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+const ImageSelectionOverlay = ({ setVisible, pushImage, numberOfImages }: Props): ReactElement => {
+  const camera = useRef<RNCamera>(null)
+
+  const takePicture = async () => {
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx:49]: Thought about this too. I think it is not correct in english to say "take a image", but i think I wasted to much thoughts on this and being consistent is better than good grammar :D

File: src/routes/process-user-vocabulary/UserVocabularyProcessScreen.tsx [hunk 9]
```diff
@@ -36,11 +46,23 @@ const StyledHintText = styled(HintText)`
...
-const UserVocabularyProcessScreen = (): ReactElement => {
+const ThumbnailContainer = styled.View`
+  flex-direction: row;
+`
+
+interface UserVocabularyProcessScreenProps {
+  navigation: StackNavigationProp<RoutesParams, 'UserVocabularyProcess'>
+}
+
+const UserVocabularyProcessScreen = ({ navigation }: UserVocabularyProcessScreenProps): ReactElement => {
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/routes/process-user-vocabulary/UserVocabularyProcessScreen.tsx:57]: Not from this PR. It was the best wording for "create and edit"-Screen we could think of. We can change this, if you want to, please add a comment in LUN-401 with a suggestion.

File: src/routes/process-user-vocabulary/UserVocabularyProcessScreen.tsx [hunk 10]
```diff
@@ -53,17 +75,56 @@ const UserVocabularyProcessScreen = (): ReactElement => {
...
+  const onSave = async (): Promise<void> => {
+    const hasError = word.length === 0 || !articleId || images.length === 0
 
-    if (!hasError) {
-      setWord('')
-      setArticleId(null)
+    if (hasError) {
+      setHasWordErrorMessage(word.length === 0)
+      setHasArticleErrorMessage(!articleId)
+      setHasImageErrorMessage(images.length === 0 || images.length > 3)
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/routes/process-user-vocabulary/UserVocabularyProcessScreen.tsx:84]: This is already implemented.

File: src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+      <Camera ref={camera} captureAudio={false} testID='camera'>
+        <Container>
+          <ActionBar>
+            {/* TODO LUN-440 implement gallery */}
+            {/* eslint-disable-next-line no-console */}
+            <PressableOpacity onPress={() => console.log('open gallery')}>
+              <ImageIcon width={GALLERY_ICON_SIZE} height={GALLERY_ICON_SIZE} testID='gallery-icon' />
+            </PressableOpacity>
+            <TakeImageButton onPress={takeImage}>
+              <CircleIconWhite width={TAKE_IMAGE_ICON_SIZE} height={TAKE_IMAGE_ICON_SIZE} testID='take-image-icon' />
```
Comment (sarahsporck) [inline src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx:67]: The feedback after taking an image is extremely bad. I didn't realise I've taken an image after I already have made three and got an error message. Will this be addressed in another issue?

**Review by sarahsporck — APPROVED** [review]
Tested on Android.

**Comment by sarahsporck** [comment]
> > 11\.
> 
> Not happend in my PR, already fixed this in LUN-420.

Created and already rejected an issue for this (LUN-444). I thought this happened because of my phone resolution ^^. Also it seems, I had a really bad timing when reviewing this PR.

File: src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx (hunk 11 above)
Comment (SteffiStuffel) [reply src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx:67]: No, this should work nicely within this ticket. As Steffen suggested, i changed the behaviour, so the cameraOverlay is closed after an image was taken and you see the thumbnail of the image as a feedback. Do you think this is a good way to handle this?

**Review by f1sh1918 — COMMENTED** [review]
Tested on android
`ListItem.tsx` line 23, pls add `flex: 1;` to fix listItem fullwidth issue
The rest steffen mentioned is fine for me since select image from gallery is a separate task. Maybe you could just comment out the gallery icon for now to not confuse testers

**Review by f1sh1918 — CHANGES_REQUESTED** [review]
Tested on android (real device) and ios (real device).
Android works fine. 
On iOS the images were not shown taken by camera. (real device & emulator)


Additionally there is an initCap for the Word missing (bar)

Looks like `document-image` couldn't be saved correctly (empty array)
ListItem should be fixed as steffen mentioned (ListItem line 23) `width: 100%;` 
(added it for my screenshot to show missing thumbnail)

File: src/routes/process-user-vocabulary/UserVocabularyProcessScreen.tsx [hunk 12]
```diff
@@ -85,33 +85,32 @@ const UserVocabularyProcessScreen = ({ navigation }: UserVocabularyProcessScreen
...
-              setArticleId(null)
-              setImages([])
-            })
-        })
+    const id = await getNextUserVocabularyId()
+    await incrementNextUserVocabularyId()
+
+    const imagePaths: Images = []
+    images.forEach(image => {
+      const index = images.indexOf(image)
```
Comment (steffenkleinle) [inline src/routes/process-user-vocabulary/UserVocabularyProcessScreen.tsx:93]: Out of curiosity, why the change from using the index from forEach to this?

File: src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx [hunk 13]
```diff
@@ -25,37 +25,31 @@ const ActionBar = styled.View`
...
 
-  const takePicture = async () => {
+  const takeImage = async () => {
     if (camera.current) {
       const options = { quality: 0.5, base64: true }
       const data = await camera.current.takePictureAsync(options)
+      setVisible(false)
       if (numberOfImages < 3) {
         pushImage(data.uri)
       }
```
Comment (steffenkleinle) [inline src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx:52]: This should not be needed anymore, right?
```suggestion
      pushImage(data.uri)
```

**Review by steffenkleinle — APPROVED** [review]
Tested on android, great job :framed_picture:

File: src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx (hunk 11 above)
Comment (sarahsporck) [reply src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx:67]: Much better now. And probably also the solution with the least amount of work. Maybe we can discuss more improvements at a later point, as they'd probably require more work. :)

**Comment by sarahsporck** [comment]
When using the article `Dropdown` I get an error saying `Closed
Virtualizedlist should never be nested inside scrollview`. This should be fixable like this: https://github.com/hossein-zare/react-native-dropdown-picker/issues/56#issuecomment-847629136.
Would you mind if you'd add this line in this PR? :)

**Comment by SteffiStuffel** [comment]
> Tested on android and ios `ListItem.tsx` line 23, pls add `flex: 1;` to fix listItem fullwidth issue 

Added with 100%, as you suggested in the other comment

>The rest steffen mentioned is fine for me since select image from gallery is a separate task. Maybe you could just comment out the gallery icon for now to not confuse testers

Won't do, as this whole feature is wip. We also didn't do this for other icons, which do nothing yet.

File: src/routes/process-user-vocabulary/UserVocabularyProcessScreen.tsx (hunk 12 above)
Comment (SteffiStuffel) [reply src/routes/process-user-vocabulary/UserVocabularyProcessScreen.tsx:93]: hmh, no idea. Undid this.

File: src/routes/process-user-vocabulary/UserVocabularyProcessScreen.tsx [hunk 14]
```diff
@@ -85,33 +85,32 @@ const UserVocabularyProcessScreen = ({ navigation }: UserVocabularyProcessScreen
...
+      audio: '',
+      alternatives: [],
+    })
+
+    navigation.navigate('UserVocabularyList', { headerBackLabel: getLabels().general.back })
+
+    setWord('')
+    setArticleId(null)
+    setImages([])
   }
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/routes/process-user-vocabulary/UserVocabularyProcessScreen.tsx:114]: Added try catch

**Comment by SteffiStuffel** [comment]
> When using the article `Dropdown` I get an error saying `Closed Virtualizedlist should never be nested inside scrollview`. This should be fixable like this: [hossein-zare/react-native-dropdown-picker#56 (comment)](https://github.com/hossein-zare/react-native-dropdown-picker/issues/56#issuecomment-847629136). Would you mind if you'd add this line in this PR? :)

Done and close the ticket @f1sh1918 opened for it.

**Review by steffenkleinle — APPROVED** [review]

**Comment by SteffiStuffel** [comment]
Tested on ios emulator and real devices again, works just fine.
